### DISCUSSION TITLE
Is Hono is a great choice for High Scale e-commerce backed? Which can have thousands of user per day

### DISCUSSION THREAD

**Discussion by a4arpon** [honojs/hono#2066] [Q&A]
I want to build a high scale-able and well performance e-commerce back-end server using Hono and Bun. 

There will be 300-400 routes 🙂 so that's why I need some guidance how to organize this app and how to maintain after initial release. 

Is hono is a great choice for that?

**Comment by yusukebe** [comment]
Hi @a4arpon 

Basically no problem, but @usualoma what do you think? The number of routes will be an issue. We could experiment, or if our existing routers can't handle it, we could build a router to handle many routes.

**Reply by yusukebe** [reply]
Btw, I've never heard of a problem caused by a large number of routes.

**Reply by usualoma** [reply]
I have not seen examples of applications with a large number of routings in production environments, but I don't think the number of routings is an issue. Even in tests with a large number of routings registered, they run fast and I have not seen any particular problems with memory usage.

Even with a large amount of routing and a large amount of middleware registration, RegExpRouter runs very fast, making hono a good choice for such applications.

```ts
const app = new Hono()

for (let i = 0; i < 1000; i++) {
  const id = Math.random().toString(36).substring(2, 15)
  app.get(`/page-${id}/:param`, (c) => {
    return c.text(`Page ${id} ${c.req.param('param')}`)
  })
}
```

**Reply by a4arpon** [reply]
The Route number is huge is because this is not an ordinary project. At this moment we have more than 600 clients regular and the clients has their own clients. NodeJS and Express seems very slow and outdated for my project. That's why I'm here to fireup my server using Hono and Bun. Everything works very well with hono just I am thinking that every time when we need to use router or use http methods so we need to init another `new Hono()` doesn't it hampers the performance ?   Cause every time we are creating new instances for the app

**Reply by yusukebe** [reply]
If you want performance, Hono and Bun are good choices, although there are few examples.

The followings are my thoughts:

Regarding `new Hono(),` each sub-app does not actually run but passes the routes and handlers to the main app through `app.route()`. This means the sub-app is just used for registration routes.

```ts
const mainApp = new Hono()

const subApp = new Hono()
subApp.get('/a', (c) => c.json('a'))

// Passing `subApp` routes and handlers to `mainApp`.
mainApp.route('/sub', subApp)
```

So, the problem here is not with the `new Hono` instance, but with the routes and handlers.

If possible, you could set up multiple applications for each group of endpoints and put a reverse proxy on the front. it can be also made with Hono. Cloudflare Workers has [the similar approach](https://developers.cloudflare.com/workers/configuration/bindings/about-service-bindings/) to make large applications.

@usualoma any thoughts?

**Reply by usualoma** [reply]
The rest is a matter of application design, so I don't want to go into too much depth, but basically I think the following approach as described by @yusukebe is appropriate.

> If possible, you could set up multiple applications for each group of endpoints and put a reverse proxy on the front. it can be also made with Hono. Cloudflare Workers has [the similar approach](https://developers.cloudflare.com/workers/configuration/bindings/about-service-bindings/) to make large applications.

One way to use sub-applications might be to dynamically create and transfer sub-applications in some cases, as shown below.

```ts
import { Hono } from 'hono'

const app = new Hono()
const subAppMap = new Map<string, Hono>()

app.get('/internal/reset/:subAppKey', (c) => {
  const subAppKey = c.req.param('subAppKey')
  const subApp = new Hono()
  subApp.get('/', (c) => c.text(`Hello from ${subAppKey}\n`))
  subAppMap.set(subAppKey, subApp)
  return c.text(`OK: ${subAppKey}\n`)
})

app.get('*', (c) => {
  const subAppKey = new URL(c.req.url).hostname
  const subApp = subAppMap.get(subAppKey)
  if (!subApp) {
    return c.text(`Not found: ${subAppKey}\n`)
  }
  return subApp.fetch(c.req.raw)
})

export default app
```

```
$ curl -H 'Host: app1' http://localhost:3000/
Not found: app1
$ curl http://localhost:3000/internal/reset/app1
OK: app1
$ curl -H 'Host: app1' http://localhost:3000/
Hello from app1
```

**Reply by movahhedi** [reply]
> Regarding `new Hono(),` each sub-app does not actually run but passes the routes and handlers to the main app through `app.route()`. This means the sub-app is just used for registration routes.
> ...
> So, the problem here is not with the `new Hono` instance, but with the routes and handlers.

I ran a little benchmark on how much it takes for `new Hono()` to run:

For me creating a Hono instance 1 million times takes 3s (`2799.66479998827 ms`).

```ts
const start = performance.now();

for (let i = 0; i < 1_000_000; i++) {
	const a = new Hono();
}

const took = performance.now() - start;
```

However routing and passing routes is not measured in this.

**Comment by a4arpon** [comment]
Thank you so much for all your support.
